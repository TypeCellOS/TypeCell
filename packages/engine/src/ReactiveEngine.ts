/* eslint-disable @typescript-eslint/no-explicit-any */
import _ from "lodash";
import { event, lifecycle } from "vscode-lib";
import { createCellEvaluator } from "./CellEvaluator.js";
import { CodeModel } from "./CodeModel.js";
import { TypeCellContext, createContext } from "./context.js";

export type ResolvedImport = {
  module: unknown;
} & lifecycle.IDisposable;

/**
 * The ReactiveEngine automatically runs models registered to it.
 * The code of the models is passed a context ($) provided by the engine.
 * This context is how the code of different models can react to each other.
 *
 * @export
 * @class ReactiveEngine
 */
export class ReactiveEngine<T extends CodeModel> extends lifecycle.Disposable {
  private disposed = false;
  public readonly observableContext = createContext<unknown>();
  private readonly registeredModels = new Set<T>();

  private readonly evaluatorCache = new Map<
    T,
    ReturnType<typeof createCellEvaluator>
  >();

  private readonly _onOutput: event.Emitter<{ model: T; output: any }> =
    this._register(new event.Emitter<{ model: T; output: any }>());

  /**
   * Raised whenever a model is (re)evaluated, with the exports by that model
   *
   * @type {event.Event<{ model: T; output: any }>}
   * @memberof Engine
   */
  public readonly onOutput: event.Event<{ model: T; output: any }> =
    this._onOutput.event;

  private readonly _onBeforeExecution: event.Emitter<{ model: T }> =
    this._register(new event.Emitter<{ model: T }>());

  /**
   * Raised before (re)evaluating a model
   *
   * @type {event.Event<{ model: T }>}
   * @memberof Engine
   */
  public readonly onBeforeExecution: event.Event<{ model: T }> =
    this._onBeforeExecution.event;

  /**
   * @param resolveImport Function (resolver) to resolve imports
   */
  constructor(
    private resolveImport: (
      module: string,
      forModel: T
    ) => Promise<ResolvedImport | undefined>,
    private debounceMillis = 100
  ) {
    super();
  }

  public registerModelProvider(compiledModelProvider: {
    onDidCreateModel: event.Event<T>;
    models: T[];
  }) {
    compiledModelProvider.models.forEach((m) => this.registerModel(m));
    this._register(
      compiledModelProvider.onDidCreateModel((m) => {
        this.registerModel(m);
      })
    );
  }

  /**
   * Register a model to the engine. After registering, the model will be observed for changes and automatically re-evaluated.
   *
   * When the model is disposed (model.dispose()), the model is automatically unregistered.
   * @param model model to register
   */
  public registerModel(model: T) {
    if (this.disposed) {
      throw new Error("registering model on disposed engine");
    }
    if (this.registeredModels.has(model)) {
      console.warn("model already registered"); // TODO: shouldn't happen
      return;
    }
    this.registeredModels.add(model);
    const evaluate = () => {
      this.evaluateUpdate(
        model,
        this.observableContext,
        async (moduleName: string) => {
          const ret = await this.resolveImport(moduleName, model);
          if (!ret) {
            throw new Error(`Could not resolve import ${moduleName}`);
          }
          this._register(ret);
          if (this.disposed) {
            // disposed in the meantime
            ret.dispose();
          }
          return ret.module;
        },
        (model, output) => this._onOutput.fire({ model, output })
      ); // catch errors?
    };
    let prevValue: string | undefined = model.getValue();

    // TODO: maybe only debounce (or increase debounce timeout) if an execution is still pending?
    const reEvaluate = _.debounce(() => {
      if (model.getValue() !== prevValue) {
        // make sure there were actual changes from the previous value

        prevValue = model.getValue();
        evaluate();
      } else {
        // TODO: inspect when this is the case. For initialization it makes sense,
        // but why do we get duplicate events more often?
      }
    }, this.debounceMillis);

    this._register(model.onDidChangeContent(reEvaluate));

    // evaluate initial
    evaluate();

    this._register(
      model.onWillDispose(() => {
        this.registeredModels.delete(model);
        const evaluator = this.evaluatorCache.get(model);
        if (evaluator) {
          evaluator.dispose();
          this.evaluatorCache.delete(model);
        }
      })
    );
  }

  public dispose() {
    if (this.disposed) {
      throw new Error("Engine already disposed");
    }
    this.disposed = true;

    super.dispose();
    this.evaluatorCache.forEach((e) => e.dispose());
  }

  private async evaluateUpdate(
    model: T,
    typecellContext: TypeCellContext<any>,
    resolveImport: (module: string) => Promise<any>,
    onOutput: (model: T, output: any) => void
  ) {
    if (!this.evaluatorCache.has(model)) {
      this.evaluatorCache.set(
        model,
        createCellEvaluator(
          typecellContext,
          resolveImport,
          true,
          (output) => onOutput(model, output),
          () => this._onBeforeExecution.fire({ model })
        )
      );
    }
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const evaluator = this.evaluatorCache.get(model)!;
    if (model.language !== "javascript") {
      throw new Error("can not evaluate non-javascript code");
    }
    const code = model.getValue();
    console.log("evaluating", code);
    await evaluator.evaluate(code);
  }
}
