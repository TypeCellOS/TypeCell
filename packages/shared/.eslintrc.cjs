module.exports = {
  root: true,
  extends: [
    "eslint:recommended",
    "plugin:@typescript-eslint/recommended",
    "react-app",
    "react-app/jest",
  ],
  parser: "@typescript-eslint/parser",
  plugins: ["import", "@typescript-eslint"],
  rules: {
    curly: 1,
    "import/no-extraneous-dependencies": [
      "error",
      {
        devDependencies: true,
        optionalDependencies: false,
        peerDependencies: false,
        bundledDependencies: false,
      },
    ],
  },
};
