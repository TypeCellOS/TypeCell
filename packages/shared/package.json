{
  "name": "@typecell-org/shared",
  "version": "0.0.3",
  "private": true,
  "dependencies": {
    "typecell-org/util": "^0.0.3"
  },
  "devDependencies": {
    "typescript": "5.0.4"
  },
  "source": "src/index.ts",
  "types": "types/src/index.d.ts",
  "main": "dist/src/index.js",
  "type": "module",
  "scripts": {
    "clean": "rimraf dist && rimraf types",
    "build": "npm run clean && tsc -p tsconfig.json",
    "watch": "tsc --watch"
  }
}
