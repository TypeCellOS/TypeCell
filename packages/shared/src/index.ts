export * from "./Ref.js";
export * from "./referenceDefinitions/child.js";
export * from "./referenceDefinitions/fork.js";
export * from "./schema.js";
