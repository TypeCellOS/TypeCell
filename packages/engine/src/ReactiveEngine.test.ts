import { CodeModel } from "@typecell-org/shared";
import { describe, expect, it } from "vitest";
import { event } from "vscode-lib";
import { ReactiveEngine } from "./ReactiveEngine.js";
import {
  buildMockedModel,
  importResolver,
  toAMDFormat,
  waitTillEvent,
} from "./tests/util/helpers.js";

/**
 * @vitest-environment jsdom
 */

const getModel1 = () =>
  buildMockedModel(
    "model1",
    `let x = 4; 
let y = 6; 
let sum = x + y;
exports.sum = sum;
exports.default = sum;`
  );

const getModel2 = () =>
  buildMockedModel("model2", `exports.default = $.sum - 5;`);

describe("ReactiveEngine class", () => {
  it("should execute a single model", async () => {
    const engine = new ReactiveEngine<CodeModel>(importResolver);
    engine.registerModel(getModel1());

    const { model, output } = await event.Event.toPromise(engine.onOutput);

    expect(model.path).toBe("model1");
    expect(output.sum).toBe(10);
    expect(output.default).toBe(10);
  });

  it("should read exported variables from other models", async () => {
    const engine = new ReactiveEngine<CodeModel>(importResolver);
    engine.registerModel(getModel1());
    await event.Event.toPromise(engine.onOutput);

    engine.registerModel(getModel2());
    const { model, output } = await event.Event.toPromise(engine.onOutput);

    expect(model.path).toBe("model2");
    expect(output.default).toBe(5);
  });

  it("should re-evaluate code after change", async () => {
    const engine = new ReactiveEngine<CodeModel>(importResolver);
    const model1 = getModel1();

    engine.registerModel(model1);
    await event.Event.toPromise(engine.onOutput);

    model1.updateCode(
      toAMDFormat(`let x = 0;
    let y = 6;
    let sum = x + y;
    exports.sum = sum;
    exports.default = sum;`)
    );

    const { output } = await event.Event.toPromise(engine.onOutput);

    expect(output.sum).toBe(6);
    expect(output.default).toBe(6);
  });

  it("should re-evaluate other models when global variable changes", async () => {
    const engine = new ReactiveEngine<CodeModel>(importResolver);
    // TODO: Expected 4 events. Figure out why model 2 re-evaluates.
    const eventsPromise = waitTillEvent(engine.onOutput, 5);
    const model1 = getModel1();
    const model2 = getModel2();

    engine.registerModel(model1);
    engine.registerModel(model2);

    model1.updateCode(
      toAMDFormat(`let x = 0;
    let y = 6;
    let sum = x + y;
    exports.sum = sum;
    exports.default = sum;`)
    );

    const events = await eventsPromise;
    const eventsSnapshot = events.map((event) => ({
      path: event.model.path,
      output: event.output,
    }));
    const finalEvent = eventsSnapshot[eventsSnapshot.length - 1];

    expect(finalEvent.path).toBe("model2");
    expect(finalEvent.output.default).toBe(1);
    expect(eventsSnapshot).toMatchSnapshot();
  });
});
