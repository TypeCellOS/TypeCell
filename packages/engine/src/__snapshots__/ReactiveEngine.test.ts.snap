// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ReactiveEngine class > should re-evaluate other models when global variable changes 1`] = `
[
  {
    "output": {
      "default": 10,
    },
    "path": "model1",
  },
  {
    "output": {
      "default": 5,
    },
    "path": "model2",
  },
  {
    "output": {
      "default": 5,
    },
    "path": "model2",
  },
  {
    "output": {
      "default": 6,
    },
    "path": "model1",
  },
  {
    "output": {
      "default": 1,
    },
    "path": "model2",
  },
]
`;

exports[`ReactiveEngine class should re-evaluate other models when global variable changes 1`] = `
Array [
  Object {
    "output": Object {
      "default": 10,
    },
    "path": "model1",
  },
  Object {
    "output": Object {
      "default": 5,
    },
    "path": "model2",
  },
  Object {
    "output": Object {
      "default": 5,
    },
    "path": "model2",
  },
  Object {
    "output": Object {
      "default": 6,
    },
    "path": "model1",
  },
  Object {
    "output": Object {
      "default": 1,
    },
    "path": "model2",
  },
]
`;
